pool/connection_pool.go

```diff
@@ -1109,8 +1119,53 @@ func (p *ConnectionPool) handlerDeactivated(conn *tarantool.Connection,
 	}
 }
 
-func (p *ConnectionPool) fillPools() bool {
+func (p *ConnectionPool) deactivateConnection(addr string,
+	conn *tarantool.Connection, role Role) {
+	p.deleteConnection(addr)
+	conn.Close()
+	p.handlerDeactivated(conn, role)
+}
+
+func (p *ConnectionPool) deactivateConnections() {
+	for address, endpoint := range p.addrs {
+		if endpoint == nil {
+			continue
+		}
+		p.deactivateConnection(address, endpoint.conn, endpoint.role)
+	}
+}
+
+func (p *ConnectionPool) processConnection(conn *tarantool.Connection,
+	addr string, end *endpoint) bool {
+	role, err := p.getConnectionRole(conn)
+	if err != nil {
+		conn.Close()
+		log.Printf("tarantool: storing connection to %s failed: %s\n", addr, err)
+		return false
+	}
+
+	if p.handlerDiscovered(conn, role) {
+		if p.addConnection(addr, conn, role) != nil {
+			conn.Close()
+			p.handlerDeactivated(conn, role)
+		}
```

Comment:
There were no return in the original code. If we call `conn.Close()`, the conn state will be changed to `connClosed`. So, `ConnectedNow()` will return false (it check if the state is equal to `connConnected`). So, the only problem that could occur, is in implementation of `Deactivate` function for `handlerDeactivated` (but the error will be `log.Printf`-ed). But again, it is strange that there were no return previously.

Reply:
But is it call `p.handlerDeactivated` twice now?

Reply:
I think we could write the function in such manner. It would be less confusing and we will not miss anything:
```Go
func (p *ConnectionPool) processConnection(conn *tarantool.Connection,
	addr string, end *endpoint) bool {
	role, err := p.getConnectionRole(conn)
	if err != nil {
		conn.Close()
		log.Printf("tarantool: storing connection to %s failed: %s\n", addr, err)
		return false
	}

	if p.handlerDiscovered(conn, role) {
		if p.addConnection(addr, conn, role) == nil {
			end.conn = conn
			end.role = role
			return true
		}
		conn.Close()
		p.handlerDeactivated(conn, role)
	} else {
		conn.Close()
	}
	return false
}
```

Reply:
Updated.